blog/2024-09-02-amqp-flow-control/index.md

```diff
@@ -0,0 +1,828 @@
+└────────────────────────┴─────────┴──────────┘
+```
+
+While the AMQP 0.9.1 client consumes only `1,000,000 - 990,932 = 9,068` messages, the AMQP 1.0 client consumes `1,000,000 - 336,229 = 663,771` messages.
+
+Therefore, in this benchmark, the AMQP 1.0 client receives 73 times (!) more messages than the AMQP 0.9.1 client.
+
+</details>
+
+:::
+
+:::note
+
+In AMQP 0.9.1, [consumer prefetch](/docs/consumer-prefetch) limits the number of unacknowledged messages.
+When a consumer acknowledges messages by sending `basic.ack` frames, RabbitMQ delivers additional messages.
+
+In AMQP 1.0, message acknowledgment is independent of link flow control.
+A consumer acknowledging messages by sending [disposition](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-disposition) frames will not prompt RabbitMQ to deliver more messages.
+Instead, the client must replenish link credit by sending `flow` frames for RabbitMQ to continue sending messages.
+For convenience, some AMQP 1.0 client libraries automatically send both `disposition` and `flow` frames when your application acknowledges messages.
+
+:::
+
+### Delivery-Count
+
+So far, we understand only one field of the [flow](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-flow) frame: `link-credit`.
+
+What happens in the following scenario?
+```
+Receiver                                  Sender
+=======================================================================
+                              ...
+link state:                               link state:
+ link-credit = 3                           link-credit = 3
+
+flow(link-credit = 6)     ---+   +---     transfer(...)
+                              \ /
+                               x
+                              / \
+                          <--+   +-->
+
+link state:                               link state:
+ link-credit = 5                           link-credit = 6
+```
+Initially, `link-credit` is 3.
+The receiver decides to send a `flow` frame, setting the new `link-credit` to 6.
+In parallel, the sender sends a `transfer` frame.
+
+Since the receiver received the `transfer` frame after it sent the `flow` frame, it will compute its new link-credit as `6 - 1 = 5`.
+However, because the sender received the `flow` frame after it sent the `transfer` frame, it will set the credit to 6.
+As a result, the state - and therefore the view of how many credits the link has left - becomes misaligned.
+This is problematic because the sender could potentially overflow the receiver.
+
+To prevent such misalignments, a second field is needed in both the link state and the `flow` frame:
+```xml
+<field name="delivery-count" type="sequence-no"/>
+```
+
+The delivery-count is decreased by 1 each time a message is transferred.
+Specifically, the sender decreases delivery-count by 1 whenever it sends a message, and the receiver decreases delivery-count by 1 whenever it receives a message.
```

Comment:
Not sure were to bring this up, hopefully here is acceptable.

Shouldn't `delivery-count` be increased/incremented? The graph below seems to increment.

Interesting blog post, thanks!

Reply:
You're right @Darsstar.
Thank you for the detailed review.
Fixed in https://github.com/rabbitmq/rabbitmq-website/pull/2159